### DISCUSSION TITLE
Passing env vars (environ) to interpreter, specifically PYTHONHASHSEED

### DISCUSSION THREAD

**Discussion by davetapley** [pyinstaller/pyinstaller#7496] [Features & Development]
This was already raised here:
- https://github.com/pyinstaller/pyinstaller/issues/4227

...but closed immediately with link to docs by @htgoebel

The linked page is gone, but from [archive.org](https://web.archive.org/web/20190729031926/https://pythonhosted.org/PyInstaller/spec-files.html) it looks the same as [this page in current docs](https://pyinstaller.org/en/stable/spec-files.html#giving-run-time-python-options),
and in neither do I see anything which suggests this can be done?

I _assume_ it hasn't been needed because of [workaround](https://stackoverflow.com/a/73534103/21115) using `--runtime-hook` and `environ`. 
Unfortunately that won't work for `PYTHONHASHSEED` because [it needs to be set **before** the interpreter starts](https://stackoverflow.com/a/32540103/21115)

More info: https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED

---

It gets worse: I'm using `multiprocessing.context.SpawnContext`, and with:
```sh
set PYTHONHASHSEED=0
python main.py
```

then the sub-processes all correctly inherit `PYTHONHASHSEED`.

But if I build and then:
```sh
set PYTHONHASHSEED=0
dist/main/main
```

then `PYTHONHASHSEED` is not honored.

I _think_ this means I'm just stuck, and there is _no way_ to disable hash randomization for subprocesses with PyInstaller? 😬 

---

Can someone more knowledgable confirm / deny?

If it is unsupported, then per #4227 how realistic is a PR to support this?

**Comment by rokm** [comment]
If setting `PYTHONHASHSEED` environment variable does not work for the frozen application, this likely means that the environment variable is parsed in the configuration parsing codepath that is not executed when PyInstaller's bootloader sets up its embedded python interpreter. 

The bootloader might need to set [Py_HashRandomizationFlag](https://docs.python.org/3/c-api/init.html#c.Py_HashRandomizationFlag), which we currently do not touch. On the other hand, the environment variable might also be ignored because we toggle [Py_IgnoreEnvironmentFlag](https://docs.python.org/3/c-api/init.html#c.Py_IgnoreEnvironmentFlag) (which happens [here](https://github.com/pyinstaller/pyinstaller/blob/77022483b440448c62efd264ab5c5d5dd26b2e03/bootloader/src/pyi_pythonlib.c#L195)).

> I think this means I'm just stuck, and there is no way to disable hash randomization for subprocesses with PyInstaller? :grimacing:

Looks like it, yeah. 

And as you noted, this pertains interpreter initialization, so it is different from the problem of passing env vars to the frozen application, which can indeed be done via custom runtime hook.

> If it is unsupported, then per https://github.com/pyinstaller/pyinstaller/issues/4227 how realistic is a PR to support this?

At a quick glance, I'd say it's non-trivial; while we could add a bootloader flag, the problem is how to toggle the setting in the interpreter's configuration:
- one way would be to set the environment variable within bootloader before starting the embedded interpreter, but that would likely require disabling the `Py_IgnoreEnvironmentFlag`, which would make the frozen application susceptible to other `PYTHON*` environment variables being set in the environment.
- the only way to directly set the hash seed seems to be via the config structure (as opposed to global config flags), which we do not currently use as we try to keep bootloader python-version-agnostic.

So the first option is not something I'd put to general use, and the second one is pretty much no-go (at least until global config flags are removed and we're forced to somehow deal with the config structure and its variants across different python versions).

If you really need hash randomization disabled and can live with potential effects of other `PYTHON*` environment variables, I'd suggest you to go with custom build of PyInstaller's bootloader that implements the first option for your application.

**Reply by davetapley** [reply]
Thanks for thorough reply 🙏🏻 

I wish `Py_HashRandomizationFlag` wasn't so goofy. [Here](https://docs.python.org/3/c-api/init.html) says:

> Set to 1 if the [PYTHONHASHSEED](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED) environment variable is set to a non-empty string.
> 
> If the flag is non-zero, read the [PYTHONHASHSEED](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED) environment variable to initialize the secret hash seed.

It _feels like_ that should also say "Set to 0 to disable hash randomization", which is sort of awkwardly suggested [here](https://docs.python.org/3/using/cmdline.html#cmdoption-R) for the `-R` switch:

> Turn on hash randomization. This option only has an effect if the [PYTHONHASHSEED](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONHASHSEED) environment variable is set to 0, since hash randomization is enabled by default.

I don't know how easy it would be to test that hypothesis? 🤔

**Reply by rokm** [reply]
Grepping for that flag in CPython sources, it seems to me that `Py_HashRandomizationFlag` is a read-only flag. I.e., it is set from the config structure's fields (`use_hash_seed` and `hash_seed`). I can't seem to find any reference of it actually controlling anything, in contrast to other `Py_*` global flags which seem to be also read into the corresponding config structure's fields.
